src/index/LocatedTriples.cpp

```diff
@@ -0,0 +1,240 @@
+  // This method should only be called if there are located triples in the
+  // specified block.
+  AD_CONTRACT_CHECK(map_.contains(blockIndex));
+
+  // TODO<qup42>: We're assuming that the index columns are always {0, 1, 2},
+  // {1, 2} or {2}. Is this true?
+  AD_CONTRACT_CHECK(numIndexColumns <= block.numColumns());
+  AD_CONTRACT_CHECK(result.numColumns() == block.numColumns());
+  AD_CONTRACT_CHECK(result.numColumns() >= 1);
+
+  auto resultEntry = result.begin() + offsetInResult;
+  const auto& locatedTriples = map_.at(blockIndex);
+  auto locatedTriple = locatedTriples.begin();
+
+  // Advance to the first located triple in the specified range.
+  auto cmpLt = [&numIndexColumns](auto lt, auto& row) {
+    if (numIndexColumns == 3) {
+      return (row[0] > lt->triple_[0] ||
+              (row[0] == lt->triple_[0] &&
+               (row[1] > lt->triple_[1] ||
+                (row[1] == lt->triple_[1] && row[2] > lt->triple_[2]))));
+    } else if (numIndexColumns == 2) {
+      return (row[0] > lt->triple_[1] ||
+              (row[0] == lt->triple_[1] && (row[1] > lt->triple_[2])));
+    } else {
+      AD_CORRECTNESS_CHECK(numIndexColumns == 1);
+      return (row[0] > lt->triple_[2]);
+    }
+  };
+  auto cmpEq = [&numIndexColumns](auto lt, auto& row) {
+    if (numIndexColumns == 3) {
+      return (row[0] == lt->triple_[0] && row[1] == lt->triple_[1] &&
+              row[2] == lt->triple_[2]);
+    } else if (numIndexColumns == 2) {
+      return (row[0] == lt->triple_[1] && row[1] == lt->triple_[2]);
+    } else {
+      AD_CORRECTNESS_CHECK(numIndexColumns == 1);
+      return (row[0] == lt->triple_[2]);
+    }
+  };
+  auto writeTripleToResult = [&numIndexColumns, &result](auto resultEntry,
+                                                         auto& locatedTriple) {
+    for (size_t i = 0; i < numIndexColumns; i++) {
+      (*resultEntry)[i] = locatedTriple->triple_[3 - numIndexColumns + i];
+    }
+    // Write UNDEF to any additional columns.
+    for (size_t i = numIndexColumns; i < result.numColumns(); i++) {
+      (*resultEntry)[i] = ValueId::makeUndefined();
+    }
+  };
+
+  for (auto row : block) {
+    // Append triples that are marked for insertion at this `rowIndex` to the
+    // result.
+    while (locatedTriple != locatedTriples.end() && cmpLt(locatedTriple, row)) {
+      if (locatedTriple->shouldTripleExist_ == true) {
+        // Adding an inserted Triple between two triples.
+        writeTripleToResult(resultEntry, locatedTriple);
+        ++resultEntry;
+        ++locatedTriple;
+      } else {
+        // Deletion of a triple that does not exist in the index has no effect.
+        ++locatedTriple;
+      }
+    }
+
+    // Append the triple at this position to the result if and only if it is not
+    // marked for deletion and matches (also skip it if it does not match).
+    bool deleteThisEntry = false;
+    if (locatedTriple != locatedTriples.end() && cmpEq(locatedTriple, row)) {
```

Comment:
The typical way to handle such remainders is to immediately break once one of the inputs is exhausted, and to then
handle the remainder at the end.